[VectorCombine] Enable transform 'scalarizeLoadExtract' for scalable vector types

benshi001 opened the pull request:
The transform 'scalarizeLoadExtract' can be applied to scalable
vector types if the index is less than the minimum number of elements.

The transform 'scalarizeLoadExtract' can be applied to scalable
vector types if the index is less than the minimum number of elements.

The check whether the index is less than the minimum number of elements
locates at line 1175~1180. 'scalarizeLoadExtract' will call 'canScalarizeAccess'
and check the returned result if this transform is safe.

At the beginning of the function 'canScalarizeAccess', the index will be checked
1. if it is less than the number of elements of a fixed vector type.
2. if it is less than the minimum number of elements of a scalable vector type.

Otherwise 'canScalarizeAccess' will return unsafe and this transform will be
prevented.

This PR is stacked https://github.com/llvm/llvm-project/pull/65442

fhahn commented:
> The transform 'scalarizeLoadExtract' can be applied to scalable vector types if the index is less than the minimum number of elements.
> 


Would be good to clarify this in the commit message; we need to prove that the index is < # of vector elements independent of whether the vector is scalable or not. IIUC the restriction on minimum number of elements is only because it makes things work well with the existing reasoning, but this restriction could be weakened in the future, i.e. there may be scenarios where we could prove an index < vscale * min number of elements.

benshi001 committed:
[VectorCombine] Enable transform 'scalarizeLoadExtract' for scalable vector types

The transform 'scalarizeLoadExtract' can be applied to scalable
vector types if the index is less than the minimum number of elements.

The check whether the index is less than the minimum number of elements
locates at line 1175~1180. 'scalarizeLoadExtract' will call 'canScalarizeAccess'
and check the returned result if this transform is safe.

At the beginning of the function 'canScalarizeAccess', the index will be checked
1. if it is less than the number of elements of a fixed vector type.
2. if it is less than the minimum number of elements of a scalable vector type.

Otherwise 'canScalarizeAccess' will return unsafe and this transform will be
prevented.

benshi001 commented:
> > The transform 'scalarizeLoadExtract' can be applied to scalable vector types if the index is less than the minimum number of elements.
> 
> Would be good to clarify this in the commit message; we need to prove that the index is < # of vector elements independent of whether the vector is scalable or not. IIUC the restriction on minimum number of elements is only because it makes things work well with the existing reasoning, but this restriction could be weakened in the future, i.e. there may be scenarios where we could prove an index < vscale * min number of elements.

I have updated my commit message for the clarification as you suggested.
https://github.com/llvm/llvm-project/pull/65443/commits/db60c013379f33c5749d2ed6448d27b1c7487779

benshi001 commented:
ping ...

nikic reviewed (approved):
LGTM

aartbik commented:
This should be fixed by https://github.com/llvm/llvm-project/pull/66670

On Mon, Sep 18, 2023 at 9:57 AM kda ***@***.***> wrote:

> Looks like this is breaking sanitizer buildbots:
> https://lab.llvm.org/buildbot/#/builders/5/builds/36746
> https://lab.llvm.org/buildbot/#/builders/74/builds/22062
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/llvm/llvm-project/pull/65443#issuecomment-1723979466>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AJPOSJ4ESZMDIFZ2FEXGOHDX3B4OTANCNFSM6AAAAAA4M25RSE>
> .
> You are receiving this because you are on a team that was mentioned.Message
> ID: ***@***.***>
>


-- 
Aart J.C. Bik, Google Inc.
https://www.aartbik.com/

aartbik commented:
Oops, wrong thread. I fixed a breakage, but not from this revision ;-)

On Mon, Sep 18, 2023 at 10:00 AM Aart Bik ***@***.***> wrote:

> This should be fixed by https://github.com/llvm/llvm-project/pull/66670
>

benshi001 commented:
> https://lab.llvm.org/buildbot/#/builders/74/builds/22062

I have fixed it by https://github.com/llvm/llvm-project/commit/87143ff9f2f4ec897f41a40c4e410c412cca697e
